File: Labs/Lab2/Lab2.py
import re
'''/**Lab2
 * Given two strings s and t, return true if t is an anagram of s, and false otherwise.
 * An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, 
 * typically using all the original letters exactly once.
 * Hint: There are 2 well-known ways to solve this problem!
 */
'''
def anagram(s, t):
    s = re.sub(r'[^a-zA-Z0-9]', '', s.lower())  # Convert to lowercase and remove non-alphanumeric characters
    t = re.sub(r'[^a-zA-Z0-9]', '', t.lower())  # Convert to lowercase and remove non-alphanumeric characters

    if len(s) != len(t):  # Check if strings have different lengths
        return False

    arrayS = [0] * len(s)  # array to hold characters of String s
    countS = [0] * 26      # alphabet count for String s
    for i in range(0, len(arrayS), 1):  # fills arrayS with char from String s
        arrayS[i] = ord(s[i]) - ord('a')  # decomposing each letter into #'s (0-25)
        countS[arrayS[i]] += 1            # stores into countS array

    arrayT = [0] * len(t)  # array to hold characters of String t
    countT = [0] * 26      # alphabet count for String t
    for j in range(0, len(arrayT), 1):  # fills arrayT with char from String t
        arrayT[j] = ord(t[j]) - ord('a')  # decomposing each letter into #'s (0-25)
        countT[arrayT[j]] += 1            # stores into countT array

    # compare countS & countT
    for k in range(0, len(arrayS), 1):
        if countS[k] != countT[k]:
            return False
    return True

#Test Cases
print("hello", "-->", "ohell", "-->",anagram("hello", "ohell")) #Expected: True
print("listen", "-->", "silent", "-->", anagram("listen", "silent")) #Expected: True
print("Extra", "-->", "Extraa", "-->", anagram("Extra", "Extraa")) #Expected: False
print("a gentleman", "-->", "elegant man", "-->", anagram("a gentleman", "elegant man")) #Expected: True
print("Clint Eastwood", "-->", "old west action", "-->", anagram("Clint Eastwood", "old west action")) #Expected: True
print("William Shakespeare", "-->", "I'll make a wise phrase", "-->", anagram("William Shakespeare", "I'll make a wise phrase")) #Expected: True
print("a decimal point", "-->", "I'm a dot in place", "-->", anagram("a decimal point", "I'm a dot in place")) #Expected: True

File: Labs/Lab2/Lab2.java
/**Lab2
 * Given two strings s and t, return true if t is an anagram of s, and false otherwise.
 * An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, 
 * typically using all the original letters exactly once.
 * Hint: There are 2 well-known ways to solve this problem!
 */
public class Lab2 {
	public static void main(String args[]) {
		
	}
	
	public boolean anagram(String s, String t) {
			s = s.toLowerCase().replaceAll("[^a-z]", "");
			t = t.toLowerCase().replaceAll("[^a-z]", "");
			//countSort s
			//countSort t
			//if s & t have same output Array, return true
			if(s.length() == t.length()) { //only computes if strings have same length
			
				int[] arrayS = new int[s.length()]; //array to hold characters of String s
				int[] countS = new int[26]; //alphabet count for String s
				for(int i = 0; i < arrayS.length; i++) { //fills arrayS with char from String s
					arrayS[i] = s.charAt(i) - 'a';       //decomposing each letter into #'s (0-25) 
													 //counts frequency of numbers from arrayS
					countS[arrayS[i]]++;				 //stores into countS array
				}
			
				int[] arrayT = new int[t.length()]; //array to hold characters of String t
				int[] countT = new int[26]; //alphabet count for String t
				for(int j = 0; j < arrayT.length; j++) { //fills arrayS with char from String t
					arrayT[j] = t.charAt(j) - 'a';       //decomposing each letter into #'s (0-25) 
													 //counts frequency of numbers from arrayT
					countT[arrayS[j]]++;				 //stores into countT array
				}
			
			//compare countS & countT
				for(int k = 0; k < arrayS.length; k++) { //loop invariant can also be bounded by arrayT, because T and S length are equal so either works
					if(countS[k] != countT[k]) {         //find first instance of mismatched numbers in countT and countS arrays
						return false;                    //immediately NOT an anagram
					}
				}
				return true;
			}
		return false; //automatic false, if s and t are different lengths
		}
	
}
